### ISSUE TITLE
`CloneIn` trait for AST nodes

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#4284]
We need to be able to clone AST nodes. But we cannot implement `Clone` on most AST types because `oxc_allocator::Box` is not `Clone`.

There is a good reason for that - `Box` stores the value it owns in arena allocator, so there is no `Box::new` method, only `Box::new_in` which takes an `&Allocator` param for the arena to allocate it in.

I propose a new trait `CloneIn`:

```rs
trait CloneIn<'alloc> {
    fn clone_in(&self, allocator: &'alloc Allocator) -> Self;
}
```

Impl for `oxc_allocator::Box`:

```rs
impl<'alloc, T> CloneIn<'alloc> for Box<'alloc, T>
where
    T: CloneIn<'alloc>,
{
    fn clone_in(&self, allocator: &'alloc Allocator) -> Self {
        Box::new_in(self.deref().clone_in(allocator), allocator)
    }
}
```

(I'm actually not sure if I have this right - maybe `'alloc` lifetime should be on `fn clone_in<'alloc>` not `CloneIn<'alloc>`)

Same as `Clone`, `CloneIn` can be implemented on any type where all its fields/variants are also `CloneIn`.

We could make a derive macro for `CloneIn` so can just `#[derive(CloneIn)]` on AST types.

NB: `clone_in` should always clone into the arena of the provided `allocator`, which *may be different* from the arena that `self` is in. i.e. you can use `clone_in` to copy objects from one arena to another.

**Comment by rzvxa** [comment]
I genuinely burst out laughing when I saw `MyTrousers`😆

**Comment by hyf0** [comment]
What's the blocker here? It seems the final form has been shown and there are no technical issues.

**Comment by hyf0** [comment]
Would it be more performant if there something like `size_hint` in `CloneIn` trait to infer allocator?

**Comment by overlookmotel** [comment]
> What's the blocker here? It seems the final form has been shown and there are no technical issues.

No blocker. We've just been working on other things. Is your need for it urgent?

> Would it be more performant if there something like `size_hint` in `CloneIn` trait to infer allocator?

Sorry I don't get you. How do you mean?

**Comment by hyf0** [comment]
> No blocker. We've just been working on other things. Is your need for it urgent?

Not sure how define urgent. But people are writing code like this

https://github.com/rolldown/rolldown/blob/b7dd0c99650d8770a990efa6f86daaa5bc00d469/crates/rolldown_plugin_dynamic_import_vars/src/clone_expr.rs

> Sorry I don't get you. How do you mean?

```rust
trait CloneIn<'new_alloc> {
    type Cloned;

    fn clone_in(&self, allocator: &'new_alloc Allocator) -> Self::Cloned;

    fn size_hint(&self) -> usize {
       0
    }
}

impl<'old_alloc, 'new_alloc> CloneIn<'new_alloc> for UnaryExpression<'old_alloc> {
    type Cloned = UnaryExpression<'new_alloc>;

    fn clone_in(&self, allocator: &'new_alloc Allocator) -> Self::Cloned {
        // sadly bumpalo doesn't have this API.
        allocator.reserve_additional_bytes(Self::size_hint());
        UnaryExpression {
            span: self.span.clone_in(allocator),
            operator: self.operator.clone_in(allocator),
            argument: self.argument.clone_in(allocator),
        }
    }
}
```

**Comment by overlookmotel** [comment]
> Not sure how define urgent. But people are writing code like this
> https://github.com/rolldown/rolldown/blob/b7dd0c99650d8770a990efa6f86daaa5bc00d469/crates/rolldown_plugin_dynamic_import_vars/src/clone_expr.rs

Oh my GOD! That's horrendous. Also, it screws up the spans.

> ```rust
> fn size_hint(&self) -> usize { 0 }
> ```

Ah now I understand. Yes, that would be good. But problem is how do you work out what the size is? In the case of `UnaryExpression`, `argument` is an `Expression` which could be `1` or could be `(() => { let loadsOfComplexLogic = foo(bar, qux); return !loadsOfComplexLogic; })()` - basically any size.

Only way to work out the size accurately would be to visit all the children of the AST node. I imagine performing that extra visit would cost more than what you save by having the size hint.

**Comment by rzvxa** [comment]
@hyf0 I saw you mentioned this issue, Just want to let you know I've picked this up and it should be expected in the next release.

**Comment by overlookmotel** [comment]
Implemented in #4732. Very nice work from the man like @rzvxa.

**Comment by Dunqing** [comment]
`BindingIdentifier`, `IdentifierReference`, and other `Identifier` have been derived `Clone` trait. So I want to ask should we use `clone_in` or `clone` for that? If we should use `clone_in` everywhere, wouldn't we have to remove `Clone`?
